forest_benchmarking/quantum_volume.py

```diff
@@ -0,0 +1,262 @@
+                                       num_circuits: int = 100, num_shots: int = 1000,
+                                       show_progress_bar: bool = False) -> int:
+    """
+    This method performs the bulk of the work in the quantum volume measurement; for the given
+    depth, num_circuits many random model circuits are generated, the heavy outputs are
+    determined from the ideal output distribution of each circuit, and a native quil
+    implementation of the model circuit output by the program generator is run on the qc. The total
+    number of sampled heavy outputs is returned.
+
+    :param qc: the quantum resource that will implement the PyQuil program for each model circuit
+    :param qubits: the qubits available in the qc for the program_generator to use.
+    :param depth: the depth (and width in num of qubits) of the model circuits
+    :param program_generator: a method which takes an abstract description of a model circuit and
+        returns a native quil program that implements that circuit. See measure_quantum_volume
+        docstring for specifics.
+    :param num_circuits: the number of random model circuits to sample at this depth; should be >100
+    :param num_shots: the number of shots to sample from each model circuit
+    :param show_progress_bar: displays a progress bar via tqdm if true.
+    :return: the number of heavy outputs sampled among all circuits generated for this depth
+    """
+    wfn_sim = NumpyWavefunctionSimulator(depth)
+
+    num_heavy = 0
+    # display progress bar using tqdm
+    for _ in tqdm(range(num_circuits), disable=not show_progress_bar):
+        # generate a simple list representation for each permutation of the depth many qubits
+        permutations = [np.random.permutation(range(depth)) for _ in range(depth)]
+
+        # generate a matrix representation of each 2q gate in the circuit
+        num_gates_per_layer = depth // 2
+        gate_list = np.array([haar_rand_unitary(4) for _ in range(depth * num_gates_per_layer)])
+        gates = np.reshape(gate_list, (depth, num_gates_per_layer, 4, 4))
+
+        # generate a PyQuil program in native quil that implements the model circuit
+        # The program should measure the output qubits in the order that is consistent with the
+        # comparison of the bitstring results to the heavy outputs given by collect_heavy_outputs
+        program = program_generator(qc, qubits, permutations, gates)
+
+        # run the program num_shots many times
+        program.wrap_in_numshots_loop(num_shots)
+        executable = qc.compiler.native_quil_to_executable(program)
+        results = qc.run(executable)
+
+        # classically simulate model circuit represented by the perms and gates for heavy outputs
+        heavy_outputs = collect_heavy_outputs(wfn_sim, permutations, gates)
+
+        # determine if each result bitstring is a heavy output, as determined from simulation
+        for result in results:
```

Comment:
I'm going to open a follow-up ticket to investigate whether this for loop can be cleverly vectorized with numpy functions to make it faster. Could be an interesting first ticket for a contributor?

Reply:
https://github.com/rigetti/forest-benchmarking/issues/34